elastic/kibana
# [Security Solution] Integrate Vanilla Unified Data Table in Timeline

## Fixes below issues

- https://github.com/elastic/security-team/issues/8726
- https://github.com/elastic/security-team/issues/7234

## Details

### Objective

The objective of this PR is to implement basic unified data table replacing the current timeline table. It is intended to behave exactly like discover table. Below are the functionalities of the timeline table that are out of scope of this PR and will be included in the follow up PR.

|Before|After|
|---|---|
|<video src="https://github.com/elastic/kibana/assets/7485038/fd8e4a42-06ed-4705-96cf-3ed0a22caa46" />|<video src="https://github.com/elastic/kibana/assets/7485038/d248856f-dcc0-4d5b-a710-b1218a4898f5" />|

This feature can be enabled with below feature flag:

```yaml
xpack.securitySolution.enableExperimental:
  - unifiedComponentsInTimelineEnabled

```

### Out of scope functionalities
- Row Renderers - https://github.com/elastic/security-team/issues/8728
- Notes / Pinned Events / Analyzer / Session View Link - https://github.com/elastic/security-team/issues/8727
- Drag / Drop from table --> data providers.
- Telemetry

### Desk Testing Guide

Below are some areas which have changes and would warrant some desk testing. 

One pre-requisite is to enable feature flag `unifiedComponentsInTimelineEnabled` which will replace traditional timeline table with the new unified timeline table.

1. **Pagination**
    - If total number of events are > sample size ( default 500 ), table footer on last page should display the notification so that user can `Load More` on-demand.
    - Page changes should be instant, because it is just client-side pagination where we download multiple pages ( 500 events ) at once.
3. **Flyouts**
    - Event Detail ( Open / Close ) - should be <s>new</s> old Expandable flyout - ( New Expandable fl
    - Host / User / IP details flyout ( Open / Close) 
4. Unified List
    - Add column by dragging & by clicking ⨁ control.
    - Remove columns by clicking on column header & by clicking on ❌ control
    - ![grafik](https://github.com/elastic/kibana/assets/7485038/45a8db43-7bb4-4439-a5da-ba8b5293158c)
    - ![grafik](https://github.com/elastic/kibana/assets/7485038/ee0e0a82-2baf-49d6-9c7e-1560ef6f71ce)
3. Full screen mode
6. Last updated date
7. Columns Order
    - Change columns order from table controls
8. Sort Order
    - timestamp
    - Any number column
    - Any string column
9. Column Actions
    - Move Right/Left
    - Sort Asc/Desc
    - Copy Column and Column Name
    - Edit Data View Field with custom label
8. Table Control - Display Options
    - <img width="517" alt="image" src="https://github.com/elastic/kibana/assets/7485038/efd4514a-20f1-4461-8686-78366bc9a611">
    - Try different row heights and Sample sizes.


## Current Observations of Unified table vis-a-vis discover UI + Issues
- [ ] Discover has custom UI on toolbar visibility when compared to vanilla unified data table
- [x] Full screen behaviour of unified data table
- [x] column width calculation is not automatic.
    - [x] Unsaved Timeline
    - [x] Saved Timeline 
- [x] Host/Network Flyout is not opening when timeline is not saved
    - [x] Flyout is not closing after it has been opened
- [ ] Row highlighting for building block -> Will be covered with Actions Column
- [x] Additional Controls 
   - [x] Row Renderer Selection
   - [x] Last updated Date
   - [x] Full Screen 
 - [ ] Table controls tab order
 - [ ] Refactor singleton ActiveTimeline class ( Inform @PhilippeOberti  )
 - [ ] Total Count not visible

### Checklist

Delete any items that are not applicable to this PR.
